File: graphs/depth_first_search.js
/**
 * Depth First Search Algorithm
 */

class Graph {
    constructor() {
        this.adj = new Map();
        this.visited = new Set();
    }

    addEdge(v, w) {
        if (!this.adj.has(v)) {
            this.adj.set(v, []);
        }
        this.adj.get(v).push(w);
    }

    DFS(v) {
        this.visited.add(v);
        console.log(v + " ");

        let neighbors = this.adj.get(v);
        if (neighbors) {
            for (let i of neighbors) {
                if (!this.visited.has(i)) {
                    this.DFS(i);
                }
            }
        }
    }
}

// Driver Code
let g = new Graph();
g.addEdge(0, 1);
g.addEdge(0, 2);
g.addEdge(1, 2);
g.addEdge(2, 0);
g.addEdge(2, 3);
g.addEdge(3, 3);

console.log("Following is Depth First Traversal (starting from vertex 2)");
g.DFS(2);

File: graphs/depth_first_search.rs
/**
 * Depth First Search Algorithm in Rust
 */

use std::collections::{HashMap, HashSet};

struct Graph {
    adj: HashMap<i32, Vec<i32>>,
}

impl Graph {
    fn new() -> Self {
        Graph {
            adj: HashMap::new(),
        }
    }

    fn add_edge(&mut self, v: i32, w: i32) {
        self.adj.entry(v).or_insert(Vec::new()).push(w);
    }

    fn dfs(&self, v: i32, visited: &mut HashSet<i32>) {
        visited.insert(v);
        print!("{} ", v);

        if let Some(neighbors) = self.adj.get(&v) {
            for &next in neighbors {
                if !visited.contains(&next) {
                    self.dfs(next, visited);
                }
            }
        }
    }
}

fn main() {
    let mut g = Graph::new();
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 2);
    g.add_edge(2, 0);
    g.add_edge(2, 3);
    g.add_edge(3, 3);

    println!("Following is Depth First Traversal (starting from vertex 2)");
    let mut visited = HashSet::new();
    g.dfs(2, &mut visited);
    println!();
}